### PR TITLE
notLeaderError should use originRaftNodeID as input parameter.

### PR DESCRIPTION
**Pull request by es-chow**
Found some odd log "replica 2:2 not leader; leader is 2:2" when running following command:

``` shell
./cockroach range split a
./cockroach range split b
```

log snippet:

``` log
I0616 21:18:01.923417    9995 range_command.go:754] range 2: new leader lease replica 2:2 21:18:00.000 +2.008s
E0616 21:18:01.923749    9995 range.go:782] error executing raft command InternalResolveIntent: replica 2:2 not leader; leader is 2:2
E0616 21:18:01.924795    9995 range.go:782] error executing raft command InternalPushTxn: replica 2:2 not leader; leader is 2:2
```

This happens when a InternalLeaderLease command from server 2 is proposed to raft before a InternalResolveIntent/InternalPushTxn command from server 1, so when applyRaftCommand(InternalResolveIntent) is executed, Leader has been changed, a NotLeaderError will be issued with following code snippet. newNotLeaderError method will use current store to find origin leader, the result is itself, so the odd log comes.

``` go
        // Verify the leader lease is held; Note that we don't require the                                                                                                      
        // leader lease when trying to grant the leader lease!                                                                                                                  
        if _, ok := args.(*proto.InternalLeaderLeaseRequest); !ok {
                if !r.verifyLeaderLease(originNodeID, header.Timestamp) {
                        err := r.newNotLeaderError()
                        reply.Header().SetGoError(err)
                        return err
                }
        }
```

### CODE REVIEW

File: storage/range.go [hunk 1]
```diff
@@ -304,11 +304,12 @@ func (r *Range) getLease() *proto.Lease {
 
 // newNotLeaderError returns a NotLeaderError intialized with the
 // replica for the last known holder of the leader lease.
-func (r *Range) newNotLeaderError() error {
+func (r *Range) newNotLeaderError(originRaftNodeID proto.RaftNodeID) error {
 	err := &proto.NotLeaderError{}
 	if l := r.getLease(); l.RaftNodeID != 0 {
-		_, err.Replica = r.Desc().FindReplica(r.rm.StoreID())
-		_, storeID := proto.DecodeRaftNodeID(proto.RaftNodeID(l.RaftNodeID))
+		_, storeID := proto.DecodeRaftNodeID(originRaftNodeID)
```
Comment (tamird) [inline storage/range.go]: this variable deserves a unique name, how about originStoreID?

**Comment by tamird** [comment]
LGTM. Can you clarify the commit message to indicate this only affects logging?

**Comment by tschottdorf** [comment]
LGTM. @tamird this technically didn't only affect logging, it meant that replicas returned different errors in that code path (though we never did anything with that information returned).

**Comment by tamird** [comment]
@es-chow can you add a test case for `applyRaftCommand` which might behave slightly differently in this error case now?

File: storage/range.go (hunk 1 above)
Comment (es-chow) [inline storage/range.go]: can be done.

**Comment by es-chow** [comment]
will try to add a test case.

**Comment by bdarnell** [comment]
LGTM.

This looks difficult to test without changes to `applyRaftCommand`. I think we need to get rid of the `return` when `verifyLeaderLease` fails, and instead handle the error more like we handle the error from `setAppliedIndex`. It's important that a command that fails due to an invalid leader lease proceed through the rest of this method, to update the applied index and the response cache. Once we write the NotLeaderError to the response cache, we can test by reading from all three replicas' response caches and making sure they are identical.

**Comment by tschottdorf** [comment]
@bdarnell the AppliedIndex is updated already (via `defer`) but the ResponseCache isn't (which should change, but doesn't seem to have any harmful implications).

For the test, it seems easy to have one that manipulates a Range's leader, calls `applyRaftCommand` directly and checks the error. A more end-to-end test is very involved, requiring us to expire the leader lease while a command is inside of Raft - that's certainly outside the scope of this PR, and doesn't add much value (at least for testing the contents of this PR). We'll want to have some tests using #837 once we have that, in which case triggering potentially inconsistent situations will be more interesting.

**Comment by es-chow** [comment]
It seems that only log message got affected as no code path make jump based on the value of NotLeaderError.Replica or store the value of NotLeaderError.Replica.

**Comment by tschottdorf** [comment]
Yes, your change is effectively a no-op. The test would serve mostly to future-proof this, though really the only thing you test at the end of the day is that you're calling `newNotLeaderError` with the correct `originNodeID`. I think the simplest thing will be if you follow my instructions above: 

> For the test, it seems easy to have one that manipulates a Range's leader, calls applyRaftCommand directly and checks the error.

**Comment by es-chow** [comment]
@bdarnell one doubt about this statement: ` It's important that a command that fails due to an invalid leader lease proceed through the rest of this method, to update the applied index and the response cache.`, shall we continue proceeding through the rest of this method include `r.executeCmd`? if the result of  `r.executeCmd` like Put is success, but a `NotLeaderError` is returned, is it confusing? And we should not update the response cache, if doing so, retry of the same command will fail again with the same `NotLeaderError`. Maybe we should just update the applied index as current implementation.

**Comment by bdarnell** [comment]
@tschottdorf Ah, I forgot about that defer. 

@es-chow: No, it shouldn't call `r.executeCmd` if there is a `NotLeaderError`. By "the rest of this method" I meant that a `NotLeaderError` should be handled in the same way as an error returned by `r.executeCmd`, including the applied index (which it turns out we are doing anyway) and response cache.

I'm not sure whether putting this in the response cache is the right thing to do or not. Most of the time when there is an error we do put it in the response cache (so if the operation is retried, it will have to get a new command ID). But in the case of a `NotLeaderError` there is nothing wrong with the request, so it could be retried on the new leader without going back to the client for a new command id. It seems that the main benefit to caching NotLeaderErrors is to be able to more easily test this change, but that's probably not worth it. 

I wouldn't worry too much about testing this change since it doesn't seem to have a meaningful effect on anything but logging.

**Comment by tschottdorf** [comment]
Quick note: I have a local wip that addresses the applyRaftCommand stuff (among other things)  so it's safe for this pr to ignore it.

**Comment by tschottdorf** [comment]
LGTM

File: storage/range.go (hunk 1 above)
Comment (tamird) [inline storage/range.go]: please address this before merging! :8ball:

**Comment by es-chow** [comment]
Modify the git log description and rebase to the latest code.

**Comment by tamird** [comment]
LGTM

**Comment by bdarnell** [comment]
LGTM

**Pull request merged**